cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+* CID garbage collection malfunction.
+* Network policy incorrectly dropping connections on KCP upgrade, when cilium-agent restarts. Related to CID duplication
+* Unhealthy cluster state that is difficult (or impossible) to recover from - related to numbers of IDs (and other Cilium custom resources) and CID duplication.
+
+
+# Design
+
+
+## Operator changes
+
+Operator will start managing CIDs, and will become the only system pod that issues mutating requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing (creating) CIDs.
+
+1) Operator requires watching the following resources to allocate CIDs:
+
+* Pods (Already watches) - To manage CIDs based on pod labels.
+* Namespaces - to manage CIDs based on namespace labels changes.
+* CIDs - to keep the desired state of CIDs. Required for:
+    * Getting the initial state on start-up.
+    * Being compatible with Agents managing CIDs simultaneously.
+    * Handling unplanned CID changes (create, update, delete) by other sources -- not Operator or Agent.
+
+2) Operator will modify the following resources:
+
+* CIDs - Create and update. If namespace or pod labels are changed, new CIDs will be created for all unique label sets that don’t match any existing identity. Unused CIDs will be cleaned up by the CID GC. In case of an update to a CID, the CID controller will update the CID back to the correct state, effectively reverting the initial change.
+
+3) CID GC:
+
+* In the initial release, CID controller will not delete CIDs. It will still rely on the legacy CID GC (the periodic GC) to deleted unused CIDs. The reason is to reduce the scope of the changes, and allow for simpler and smaller steps in the direction of CID controller fully managing CIDs.
+* In the next phase, CID GC will be moved to CID controller. The benefit is that CID controller will already have information about which CIDs are used and which not, and will be able to delete CIDs in real time, without periodically running GC.
+
+## Agent changes
+
+Agents will stop creating CIDs. Agents will only generate local security identities. Agents will no longer perform following actions:
+
+* Locally create and reserve security identities in the global range [256, 65536].
+* Create CIDs.
+* Remove heartbeat status annotation on CIDs (part of CID GC) - In the second phase when CID GC is moved to CID controller.
+* Agents will no longer need the permission to modify CIDs - In the second phase when CID GC is moved to CID controller.
+
+Agents will wait for CIDs to be received through the watcher, and then allocated them to local endpoints (pods) by retrieving the CIDs from the watcher's store based on pod and namespace labels.
+
+**Namespace watch**
+
+Agents will continue to watch namespaces in order to instantly update security identity labels in the local Endpoint manager’s cache. More info below in Label changes: Namespace.
```

Comment:
I'm curious, why is this necessary? The later discussion states that the Agents will wait until the Operator creates new CIDs anyway.

Reply:
This is only for the case of namespace label changes. Agents anyway already watch namespaces and use the Namespace watcher cache (Store) to retrieve namespace labels. Agents need to know what are the namespace labels for each pod, to be able to match pod+namespace labels to a CID from CID watcher's cache.

Reply:
OK. Reading the current text, it sounded like the motivation is to "instantly update", which implies that if this namespace watcher was removed then the update would just be delayed. However, this response from you implies that this step is required for correct operations:

> Agents need to know what are the namespace labels for each pod, to be able to match pod+namespace labels to a CID from CID watcher's cache.

I'd suggest merging this motivation into this section so future readers can understand that this is a required aspect of the design, not optional.

Reply:
Added.